src/aspire/operators/diag_matrix.py

```diff
@@ -0,0 +1,622 @@
+                ind += i
+
+        elif isinstance(other, np.ndarray):
+            # For now, lets just interop with 2D `other` arrays,
+            # assumed to represent matrices.
+            if other.ndim != 2:
+                raise ValueError(
+                    f"DiagMatrix.matmul of ndarray only supports 2D, received {other.shape}."
+                )
+
+            # Then the mathematical `DA` for diag D and matrix A,
+            # is expressed in code as `D@A`.
+            #
+            # The operation `DA` is known as Left Scaling,
+            # and means to scale the row A_i by d_i in D.
+            # This can be accomplished by broadcasting an elemental multiply.
+
+            res = self._data[..., np.newaxis] * other
+        else:
+            raise RuntimeError(f"__matmul__ not implemented for {type(other)}.")
+
+        return res
+
+    def __matmul__(self, other):
+        """
+        Operator overload for matrix multiply of `DiagMatrix` instances.
+        """
+
+        return self.matmul(other)
+
+    def __rmatmul__(self, lhs):
+        """
+        Compute the right matrix multiplication with a `DiagMatrix` instance,
+        and a numpy array, `lhs` @ `self`.
+
+        :param other: The lhs Numpy instance.
+        :return: Returns numpy array representing `other @ self`.
+        """
+
+        # Note, we should only hit this method when mixing DiagMatrix with numpy.
+        #   This is because if both a and b are DiagMatrix,
+        #   then a@b would be handled first by a.__matmul__(b), never reaching here.
+        #   If a is BlkDiagMatrix, then it will handle the conversion of DiagMatrix b.
+
+        if not isinstance(lhs, np.ndarray):
+            raise RuntimeError(f"__rmatmul__ not implemented for {type(lhs)}.")
+
+        # For now, lets just interop with 2D `other` arrays,
+        # assumed to represent matrices.
+        if lhs.ndim != 2:
+            raise ValueError(
+                f"DiagMatrix.matmul of ndarray only supports 2D, received {lhs.shape}."
+            )
+
+        # Then the mathematical `AD` for diag D and matrix A,
+        # is expressed in code as `A@D`.
+        #
+        # The operation `AD` is known as Right Scaling,
+        # and means to scale the column A_j by d_j in D.
+        # This can be accomplished by broadcasting an elemental multiply.
+        # In this case A is the variable `lhs`
+
+        return lhs * self._data[np.newaxis]
+
+    def mul(self, other):
+        """
+        Compute the elementwise multiplication of a `DiagMatrix` instance and a
+        scalar or another `DiagMatrix`.
+
+        :param other: The rhs in the multiplication..
+        :return: A `self` * `other` as type of `other`.
+        """
+        if isinstance(other, DiagMatrix):
+            res = DiagMatrix(self._data * other._data)
+        elif isinstance(other, np.ndarray):
+            res = self * DiagMatrix(other)
```

Comment:
I don't think this is consistent with the `BlkDiagMatrix` version below. The equivalent multiplication operation given an `ndarray` would extract its diagonal elements before performing the multiplication (the others would be multiplied by zero).

Reply:
Unless I confused myself, I think that depends if your ndarray represents a collection of vectors (stack of diags) or a matrix.

I think this was intended to handle the first, a stack of vectors.  If `other` was a matrix the operation should be some variant of `matmul` or `@`.

I'm not sure it is possible to disambiguate the two 2d arrays.   I'll think this over again after I hit the easy remarks. Thanks

Reply:
I think we want the matrix behavior. This thing is a matrix (it's in the name) and it acts like a matrix when we do matmuls, so why should it act as a vector when we do elementwise multiplication? The behavior that you're describing is obtained by casting the rhs to a `DiagMatrix`.

The way I see this class is as a way of optimizing various operations related to diagonal matrices (just `BlkDiagMatrix`) but in a way that abstracts away the underlying mechanism (working with vectors). This would suggest that it act like a matrix when adding, multiplying, etc.

On the other hand, `BlkDiagMatrix` only allows multiplication scalars, so maybe that is the way to do?

Reply:
I agree it can be obtained by casting to `DiagMatrix`. I guess I was just trying to be flexible.  (along lines of 2d np arrays have both elementwise and matmul exposed.)

Removed and raise when we are not a scalar or `DiagMatrix`.  Not a problem :) thanks.
Also updated tests.

Reply:
Great! Thanks.